lib/trino-hive-formats/src/main/java/io/trino/hive/formats/avro/NativeLogicalTypesAvroTypeManager.java

```diff
@@ -191,7 +210,99 @@ private static Function<LogicalType, BiConsumer<BlockBuilder, Object>> getLogica
         };
     }
 
-    private <T> Optional<T> validateAndProduceFromName(Schema schema, Function<LogicalType, T> produce)
+    private static BiFunction<Block, Integer, Object> getAvroFunction(LogicalType logicalType, Schema schema, Type type)
+            throws AvroTypeException
+    {
+        return switch (logicalType.getName()) {
+            case TIMESTAMP_MILLIS -> {
+                if (!(type instanceof TimestampType timestampType)) {
+                    throw new AvroTypeException("Can't represent avro logical type %s with Trino Type %s".formatted(logicalType.getName(), type));
+                }
+                if (timestampType.getPrecision() > 3) {
+                    throw new AvroTypeException("Can't write out Avro logical millisecond timestamp from Trino timestamp type with precision %s".formatted(timestampType.getPrecision()));
+                }
+                if (timestampType.isShort()) {
+                    yield (block, integer) -> timestampType.getLong(block, integer) / Timestamps.MICROSECONDS_PER_MILLISECOND;
+                }
+                else {
+                    // todo this is dead or I remove precision check above
+                    // drop precision or throw type error?
+                    yield ((block, integer) ->
+                    {
+                        SqlTimestamp timestamp = (SqlTimestamp) timestampType.getObject(block, integer);
+                        return timestamp.roundTo(3).getMillis();
+                    });
+                }
+            }
+            case TIMESTAMP_MICROS -> {
+                if (!(type instanceof TimestampType timestampType)) {
+                    throw new AvroTypeException("Can't represent avro logical type %s with Trino Type %s".formatted(logicalType.getName(), type));
+                }
+                if (timestampType.getPrecision() > 6) {
+                    throw new AvroTypeException("Can't write out Avro logical microsecond timestamp from Trino timestamp type with precision %s".formatted(timestampType.getPrecision()));
+                }
+                if (timestampType.isShort()) {
+                    // use non lambda because lambda throws exception in error prone
```

Comment:
Did you mean "method reference" instead of "lambda"? What is the failure from Error Prone?

Reply:
https://github.com/google/error-prone/issues/3991 this is the issue I filed with stack trace